component/org.wso2.carbon.identity.oauth2.grant.token.exchange/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/utils/TokenExchangeUtils.java

```diff
@@ -349,18 +396,78 @@ public static void setAuthorizedUser(OAuthTokenReqMessageContext tokenReqMsgCtx,
             } else {
                 try {
                     authenticatedUser.setUserId(accessTokenDO.getAuthzUser().getUserId());
+                    authenticatedUser.setFederatedIdPName(null);
                 } catch (UserIdNotFoundException e) {
                     handleException("Error while getting user id from the access token data object.", e);
                 }
             }
-        } else {
-            authenticatedUser.setFederatedUser(true);
-            authenticatedUser.setFederatedIdPName(identityProvider.getIdentityProviderName());
         }
+
+
         tokenReqMsgCtx.setAuthorizedUser(authenticatedUser);
         populateIdPGroupsAttribute(tokenReqMsgCtx, identityProvider, claimsSet);
     }
 
+    private static ServiceProvider getServiceProvider(OAuthTokenReqMessageContext tokenReqMsgCtx) throws IdentityOAuth2Exception {
+
+        ServiceProvider serviceProvider;
+        OAuthAppDO oAuthAppBean = (OAuthAppDO) tokenReqMsgCtx.getProperty("OAuthAppDO");
+        String tenantDomain = tokenReqMsgCtx.getOauth2AccessTokenReqDTO().getTenantDomain();
+        if (StringUtils.isEmpty(tenantDomain)) {
+            tenantDomain = MultitenantConstants.SUPER_TENANT_DOMAIN_NAME;
+        }
+
+        try {
+            ApplicationManagementService appMgtService = TokenExchangeComponentServiceHolder.getInstance()
+                    .getApplicationManagementService();
+            serviceProvider = appMgtService.getServiceProvider(oAuthAppBean.getApplicationName(), tenantDomain);
+        } catch (IdentityApplicationManagementException e) {
+            throw new IdentityOAuth2Exception(e.getMessage(), e);
+        }
+
+        if (serviceProvider == null) {
+            throw new IdentityOAuth2Exception("Error while retrieving service provider for application name: " +
+                    oAuthAppBean.getApplicationName());
+        }
+
+        return serviceProvider;
+    }
+
+    private static boolean isUserAssociated(User user, IdentityProvider idp) throws IdentityOAuth2Exception {
```

Comment:
should also check for the specific subject in idp. IIRC, there is no restriction to have multiple associations from a single IDP to a single local account. Eg. local user 'alice' may have 'google' idp associtions as 'alice@gmail.com' and 'alice123@gmail.com'

Reply:
Addressed with ec674ae